src/main/java/org/spongepowered/api/CatalogKey.java

```diff
@@ -24,10 +24,32 @@
  */
 package org.spongepowered.api;
 
+import ninja.leaping.configurate.ConfigurationNode;
+import org.spongepowered.api.data.DataSerializable;
+import org.spongepowered.api.data.manipulator.DataManipulator;
+import org.spongepowered.api.data.persistence.DataTranslator;
+import org.spongepowered.api.plugin.PluginContainer;
+import org.spongepowered.api.util.ResettableBuilder;
+
 /**
- * A catalog key.
+ * An object representation of a location or pointer to a {@link CatalogType}
+ * that can be used to retrieve said type from the {@link GameRegistry}. The
+ * key can be represented as a {@link String} by {@link #toString()}. The key
+ * is built with two parts:
+ * <ol>
+ *     <li>The Namespace</li>
+ *     <li>The Value</li>
+ * </ol>
+ * Normally, the namespace is lowercased and likewise, so is the value. This
```

Comment:
The namespace and value **must** be lowercase - we throw if they aren't. Valid characters for a namespace are `a-z0-9_.-`, valid characters for a value are `a-z0-9/._-`.

Reply:
Do we want to allow the characters `/.` in ids? Atm. I am replacing all `/`'s for advancement ids with `_`'s. As for `.`'s, they are atm only used for sound ids and can also be replaced by `_`'s (to reflect `SoundTypes` field names)?
Or is this something that needs to change for 1.13?

Reply:
But `/` and `.` are allowed to be a part of a `ResourcePath`, which will be added by a separate PR for 1.13.

Reply:
Err, what? `CatalogKey` is equal to `ResourceLocation` internally - why do we need a separate `ResourcePath`?

Reply:
@Cybermaxke I've implemented `CatalogKey` over `ResourceLocation` currently as it works quite well. The resolution of id's is also defaulted on `ResourceLocation`, so having that handling already is much better than having yet another object mapping between the two classes.

Currently implementation is *almost passing*.